### ISSUE TITLE
GHC-58481 can be extended with an example for PatternSynonyms 

### ISSUE THREAD

**Issue by Bodigrim** [haskellfoundation/error-message-index#547]
```haskell
pattern Empty <- (null -> True) where
  Empty = mempty
```

currently fails with 

```
src/Data/Text.hs:518:15: error: [GHC-58481]
    parse error on input ‘<-’
    Suggested fix: Possibly caused by a missing 'do'?
    |
518 | pattern Empty <- (null -> True) where
    |
```

which is completely misleading. It could be difficult to improve it on GHC side, but at least the error index can include such example to suggest enabling `{-# LANGUAGE PatternSynonyms #-}`.

**Comment by BinderDavid** [comment]
I just added the example in https://github.com/haskellfoundation/error-message-index/commit/4ed715732617e94680d0569a3503880cbbd9a291 I simplified the example so that it doesn't also need View Patterns to fix the example.

I guess the reason why it is difficult to fix on the GHC side is that `pattern` is not a keyword if `PatternSynonyms` is not enabled, and we therefore try to parse it as the beginning of an ordinary declaration?

(I accidentally directly pushed to main, since branch protection was disabled for admins. That should be fixed now.)
